HIVE-27234: Iceberg: CREATE BRANCH SQL implementation

zhangbutao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR refers to spark-sql  about iceberg branch ddl implementation https://github.com/apache/iceberg/pull/6617
If someone has different opinions about the sql syntax of branch, we can discuss here.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Personally, branch is more useful than snapshot in iceberg, and it is more friendly to users. We can use branch do lots of meaningfull work. 

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Added a new sql syntax and hive users can create iceberg branch using the sql.
```
ALTER TABLE tableName
{CREATE BRANCH branchName [FOR SYSTEM_VERSION AS OF {snapshotId} | FOR SYSTEM_TIME AS OF {timestamp}]
[RETAIN interval {DAYS | HOURS | MINUTES}]
[WITH SNAPSHOT RETENTION {[num_snapshots SNAPSHOTS] [interval {DAYS | HOURS | MINUTES}]}]}]
```

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
UT

TuroczyX commented:
uhhhhh! This is super cool idea!  I want it! Actually with this it is possible to play on dataset without any risk. I love it! I want it!
@deniskuzZ @ayushtkn @simhadri-g Could you please make this review a priority?

Documents about it:
https://docs.google.com/document/d/1tbATFPrKF3vNlzkgZQdaW8CAJmbjvryfrlg6C2Ci_aA/edit#heading=h.v8gsu2fe19q2

ayushtkn commented:
Didn't check the code, But it was something I was also reading coincidently 2-3 hours before only :) 

The main thing to chase is not creating a branch. But to insert into those branches. @zhangbutao We should chase that once we have the create in. Will review this in a day or two along with Denys. :) 


Just FYI. I think Iceberg has options to create Tags as well now, and an option to do cherry-pick to branches as well. They are gonna release it for spark in 1.12.0

Good to have stuff!!!

TuroczyX commented:
Tags!!!! 😍

TuroczyX commented:
![image](https://user-images.githubusercontent.com/7687174/233213641-1439d6c1-b799-4939-830a-177b7f733f86.png)

TuroczyX commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/branch/create/AlterTableCreateBranchAnalyzer.java:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  @Override
+  protected void analyzeCommand(TableName tableName, Map<String, String> partitionSpec, ASTNode command)
```
I know other does not use it, in the code, but in defensive code, when you pass a reference it should be checked if it not a null in the beginning of the method. Just to avoid any unpredicted null reference exception. It is pretty optional, because other commands does not do it at all. 

Btw nice code.

zhangbutao replied:
Thanks for comment @TuroczyX . IIUC, I think no need check null here as the parser moduel will do this.  We have defined all required and optional parameters in [`AlterClauseParser.g` ](https://github.com/apache/hive/pull/4216/files#diff-3b0a324eebcad6e32a1bf8bf42499df0739207dbe96f32a3d5e64a44eb002d0fR473), and the then Antlr parser will check if the parameter is correct.

zhangbutao commented:
@TuroczyX @ayushtkn yes, branch and tag are important feats in iceberg 1.2.0.  I have created a umbrella ticket https://issues.apache.org/jira/browse/HIVE-27233 to track the new feats about branch&tag.

> The main thing to chase is not creating a branch. But to insert into those branches.

I absolutely agree. I am exploring how to achieve this in hive. Next maybe we shoud consider upgrade iceberg to latest version 1.2.1.  In addition, we can refer to this PR to implement the sql syntax for` insert into branch`

zhangbutao commented:
https://github.com/apache/hive/pull/4252 Upgrade iceberg to 1.2.1 in order to better integrate branch&tag features.

TuroczyX commented:
yes, let's upgrade the iceberg version

ayushtkn commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -676,6 +678,32 @@ public void executeOperation(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+
+    String branchName = createBranchSpec.getBranchName();
+    ManageSnapshots manageSnapshots = icebergTable.manageSnapshots();
+    Long snapShotId = Optional.ofNullable(createBranchSpec.getSnapshotId())
+        .orElse(icebergTable.currentSnapshot().snapshotId());
```
What happens to this in case of any empty table? Just a create iceberg table and nothing. As far as I remember in that case currentSnapshot is null? Just give a check once.

zhangbutao replied:
Nice catch! fixed. Thanks very much!

ayushtkn commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -676,6 +678,32 @@ public void executeOperation(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+        .orElse(icebergTable.currentSnapshot().snapshotId());
+    LOG.info("Creating branch {} on iceberg table {}.{}", branchName, hmsTable.getDbName(),
+        hmsTable.getTableName());
+    manageSnapshots.createBranch(branchName, snapShotId);
```
nit:
snapShotId -> snapshotId

zhangbutao replied:
fixed.

ayushtkn commented on iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergBranchOperation.java:
```diff
@@ -0,0 +1,141 @@
+
+public class TestHiveIcebergBranchOperation extends HiveIcebergStorageHandlerWithEngineBase {
+
+  @Test
```
Do you need so many tests? Just create a table once and can create multiple branches of that with different parameters ?

zhangbutao replied:
I think multiple small tests are better for reading. But if you think one test is ok, I'm fine with that. ;)

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/branch/create/AlterTableCreateBranchAnalyzer.java:
```diff
@@ -0,0 +1,89 @@
+      throws SemanticException {
+    Table table = getTable(tableName);
+    validateAlterTableType(table, AlterTableType.CREATEBRANCH, false);
+    if (!"ICEBERG".equalsIgnoreCase(table.getParameters().get("table_type"))) {
```
Change to 
```
    if (!HiveMetaHook.ICEBERG.equalsIgnoreCase(table.getParameters().get(HiveMetaHook.TABLE_TYPE))) {
```

zhangbutao replied:
fixed. Thx

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/branch/create/AlterTableCreateBranchAnalyzer.java:
```diff
@@ -0,0 +1,89 @@
+      case HiveParser.TOK_RETAIN:
+        String maxRefAge = childNode.getChild(0).getText();
+        String timeUnitOfBranchRetain = childNode.getChild(1).getText();
+        maxRefAgeMs = TimeUnit.valueOf(timeUnitOfBranchRetain.toUpperCase(Locale.ENGLISH)).toMillis(Long.valueOf(maxRefAge));
```
you need primitive long only, Long.valueOf does a parseLong to get a primitive and then converts to Long, and here you would be converting back to long, because toMillis to take primitive long only.

Use ``Long.parseLong(maxRefAge)``

zhangbutao replied:
fixed. Thx.

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/branch/create/AlterTableCreateBranchAnalyzer.java:
```diff
@@ -0,0 +1,89 @@
+        if (childNode.getChildren().size() > 1) {
+          String maxSnapshotAge = childNode.getChild(1).getText();
+          String timeUnitOfSnapshotsRetention = childNode.getChild(2).getText();
+          maxSnapshotAgeMs = TimeUnit.valueOf(timeUnitOfSnapshotsRetention.toUpperCase(Locale.ENGLISH)).toMillis(Long.valueOf(maxSnapshotAge));
```
use. ```Long.parseLong(maxSnapshotAge)``

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/branch/create/AlterTableCreateBranchAnalyzer.java:
```diff
@@ -0,0 +1,89 @@
+      }
+    }
+
+    AlterTableCreateBranchSpec spec = new AlterTableCreateBranchSpec(branchName, snapshotId, maxRefAgeMs, minSnapshotsToKeep, maxSnapshotAgeMs);
```
line length should be less than 120 in hive

zhangbutao replied:
fixed. Thx.

ayushtkn reviewed (commented):
Thanx @zhangbutao, overall changes looks good, dropped some comments.

Additionally your code ain't formatted according to hive formatting rules, line length should be less than 120 and other, can refer here: https://github.com/apache/hive/blob/master/dev-support/eclipse-styles.xml

Additionally add a q test.
Where you do an explain of the create branch statement before creating the branch.

Add some failure test cases:
Branch already exist and you try to create a branch
Attempt to create a branch on a non iceberg table

In the positive test case check for the data as well.

zhangbutao commented:
@ayushtkn Thanks for review! I will probably fix these comments after a few days of vacation.

zhangbutao commented:
> Thanx @zhangbutao, overall changes looks good, dropped some comments.
> 
> Additionally your code ain't formatted according to hive formatting rules, line length should be less than 120 and other, can refer here: https://github.com/apache/hive/blob/master/dev-support/eclipse-styles.xml
> 
> Additionally add a q test. Where you do an explain of the create branch statement before creating the branch.
> 
> Add some failure test cases: Branch already exist and you try to create a branch Attempt to create a branch on a non iceberg table
> 
> In the positive test case check for the data as well.

I think i have addressed these comments. Please take a look again if you have time. Thanks @ayushtkn

zhangbutao commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -676,6 +678,35 @@ public void executeOperation(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+    Table icebergTable = IcebergTableUtil.getTable(conf, tableDesc.getProperties());
+
+    String branchName = createBranchSpec.getBranchName();
+    Optional.ofNullable(icebergTable.currentSnapshot()).orElseThrow(() -> new UnsupportedOperationException(
```
Here give a check if table have current snapshot.

zhangbutao commented on iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergBranchOperation.java:
```diff
@@ -0,0 +1,166 @@
+
+    // creating a branch which is already exists will fail
+    try {
+      shell.executeStatement(String.format("ALTER TABLE customers CREATE BRANCH %s", branchName));
```
Test: creating a branch which is already exists will fail

zhangbutao commented on iceberg/iceberg-handler/src/test/queries/negative/alter_table_create_branch_negative.q:
```diff
@@ -0,0 +1,3 @@
+create table ice_tbl (id int, name string) Stored by Iceberg;
+
+alter table ice_tbl create branch test_branch_1;
```
Qtest: creating branch on table without current snapshot will fail

zhangbutao commented on iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergBranchOperation.java:
```diff
@@ -0,0 +1,166 @@
+
+    String branchName = "test_branch_1";
+    try {
+      shell.executeStatement(String.format("ALTER TABLE nonice_tbl CREATE BRANCH %s", branchName));
```
Test: creating branch on a non-icebeg table will fail

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/parse/AlterTableCreateBranchSpec.java:
```diff
@@ -0,0 +1,63 @@
+
+import com.google.common.base.MoreObjects;
+
+public class AlterTableCreateBranchSpec {
```
can we reuse `AlterTableExecuteSpec` and define inner `CreateBranchSpec`, similar to `RollbackSpec`?

zhangbutao replied:
That's what I thought before. Also I want to reuse method: `HiveIcebergStorageHandler::executeOperation` :
https://github.com/apache/hive/blob/e413b445e95e7c8beb4e43f0f6b08a98934e1c2c/iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java#L645

But it  seems to be that these codes are for  syntax `alter table execute`.  Do you think we can reuse them for `alter table create branch
` or `alter table drop branch`?

zhangbutao replied:
Finally, i decide not to reuse codes about `alter table execute` syntax , as i think it is more important to make code clear and readable.
 I  used `AlterTableBranchSpec ` which looks like the pattern `AlterTableExecuteSpec` , and then define inner `CreateBranchSpec` which type is `CREATE_BRANCH`. We can continue to add other brancn operation e.g `DROP_BRANCH`, `REPLACE_BRANCH`.

Please take a look again to see if it is make more sense. Thanks.

deniskuzZ replied:
👍

deniskuzZ reviewed (commented):
LGTM, some minor comments
btw, should we support 'drop branch' operation?

ayushtkn commented on parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g:
```diff
@@ -477,6 +478,34 @@ alterStatementSuffixExecute
+    ;
+
+snapshotIdOfBranch
+@init { gParent.pushMsg("alter table create branch as of version", state); }
```
Makes sense to be in sync. @zhangbutao can you change it

zhangbutao replied:
My original thought was to be consistent with spark-iceberg syntax. But it seems more reasonable to be sync with current hive` timet_travel` syntax.
I will be change the syntax and add the syntax `FOR SYSTEM_TIME AS OF`.

The new syntax will be as follows:

```
ALTER TABLE tableName
{CREATE BRANCH branchName [FOR SYSTEM_VERSION AS OF {snapshotId} | FOR SYSTEM_TIME AS OF {timestamp}]
[RETAIN interval {DAYS | HOURS | MINUTES}]
[WITH SNAPSHOT RETENTION {[num_snapshots SNAPSHOTS] [interval {DAYS | HOURS | MINUTES}]}]}]
```

zhangbutao replied:
done

deniskuzZ replied:
👍

zhangbutao commented on parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g:
```diff
@@ -477,6 +478,34 @@ alterStatementSuffixExecute
+@init { gParent.pushMsg("alter table create branch as of version", state); }
+@after { gParent.popMsg(state); }
+    : KW_AS KW_OF KW_VERSION snapshotId=Number
+    -> ^(TOK_AS_OF_VERSION_BRANCH $snapshotId)
```
sure, as mentioned above, we can reuse `TOK_AS_OF_VERSION` if we use `FOR SYSTEM_VERSION AS OF` instead of `AS OF VERSION`.

zhangbutao replied:
done

zhangbutao commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/branch/create/AlterTableCreateBranchAnalyzer.java:
```diff
@@ -0,0 +1,93 @@
+      throws SemanticException {
+    Table table = getTable(tableName);
+    validateAlterTableType(table, AlterTableType.CREATEBRANCH, false);
+    if (!HiveMetaHook.ICEBERG.equalsIgnoreCase(table.getParameters().get(HiveMetaHook.TABLE_TYPE))) {
```
Sorry, can you elaborate on where this logic shoud move into? This is just a simple iceberg check  statement.

zhangbutao replied:
Thanks for your tip. Added a helper method:  `DDLUtils::validateTableIsIceberg`

zhangbutao commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -676,6 +678,35 @@ public void executeOperation(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+    LOG.info("Creating branch {} on iceberg table {}.{}", branchName, hmsTable.getDbName(),
+        hmsTable.getTableName());
+    ManageSnapshots manageSnapshots = icebergTable.manageSnapshots();
+    manageSnapshots.createBranch(branchName, snapshotId);
```
If user omits ` snapshotId ` when creating branch, then current snapshotId will be used.
if user passes a non-existent `snapshotId` when creating branch, exception `unknown snapshot` will be thrown from iceberg lib.

deniskuzZ replied:
👍

zhangbutao commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AlterTableType.java:
```diff
@@ -40,6 +40,7 @@ public enum AlterTableType {
   ALTERPARTITION("alter partition"), // Note: this is never used in AlterTableDesc.
   SETPARTITIONSPEC("set partition spec"),
   EXECUTE("execute"),
+  CREATEBRANCH("create branch"),
```
done

zhangbutao commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -676,6 +678,35 @@ public void executeOperation(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+            branchName, hmsTable.getDbName(), hmsTable.getTableName())));
+    Long snapshotId = Optional.ofNullable(createBranchSpec.getSnapshotId())
+        .orElse(icebergTable.currentSnapshot().snapshotId());
+    LOG.info("Creating branch {} on iceberg table {}.{}", branchName, hmsTable.getDbName(),
```
Added `snapshotId `. I think INFO is enough as this DDL is rarely executed.

zhangbutao commented on ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:
```diff
@@ -6709,6 +6710,15 @@ public void alterTableExecuteOperation(Table table, AlterTableExecuteSpec execut
     }
   }
 
+  public void alterTableCreateBranchOperation(Table table, AlterTableCreateBranchSpec createBranchSpec) throws HiveException {
```
I have rename it to 'alterTableBranchOperation' which can be used to create/drop/replace branch. Is it ok?

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/DDLUtils.java:
```diff
@@ -217,4 +218,11 @@ public static void setColumnsAndStorePartitionTransformSpecOfTable(
     }
   }
+
+  public static void validateTableIsIceberg(org.apache.hadoop.hive.ql.metadata.Table table)
```
doesn't it duplicate `HiveTableOperations.validateTableIsIceberg`, could we keep just one version of it 
````
  static void validateTableIsIceberg(Table table, String fullName) {
    String tableType = table.getParameters().get(TABLE_TYPE_PROP);
    NoSuchIcebergTableException.check(tableType != null && tableType.equalsIgnoreCase(ICEBERG_TABLE_TYPE_VALUE),
        "Not an iceberg table: %s (type=%s)", fullName, tableType);
  }
````

zhangbutao replied:
yep, it duplicates `HiveTableOperations.validateTableIsIceberg` as DDLTask in`ql `  package can not import iceberg class.
 
I changed the thrown exception to` Not an iceberg table: %s (type=%s)` , this means the same thing as `HiveTableOperations.validateTableIsIceberg`

zhangbutao commented:
Git rebase to fix code conflicts.

zhangbutao commented:
@deniskuzZ @ayushtkn Can we merge this PR first? Thanks.
